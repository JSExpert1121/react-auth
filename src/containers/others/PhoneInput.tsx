/**
 * @name PhoneInput
 * @description verified phone input
 * @version 1.0 (basic)
 * @author Darko
 */
import * as React from 'react';

import Box from '@material-ui/core/Box';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { makeStyles, Theme } from '@material-ui/core/styles';

// 3rd party libraries
import MuiPhoneInput from 'material-ui-phone-number';
import { MessageDescriptor, injectIntl, IntlShape } from 'react-intl';
import authMsgs from 'translations/messages/page.auth';

// apis & constants
import * as MemberApi from 'service/member';
import { SUPPORTED_COUNTRIES } from 'config/app.config';

const useStyles = makeStyles((theme: Theme) => ({
    root: {
        width: '100%'
    },
    row: {
        paddingTop: theme.spacing(2),
        width: '100%'
    },
    countryCode: {
        marginTop: theme.spacing(1)
    },
    getcode: {
        height: '100%'
    },
}));

interface IPhoneInputProps {
    verify: (country: string, phone: string, flag: boolean) => void;
    setBusy: (flag: boolean) => void;
    intl: IntlShape;
    phoneNo: string;
    countryCode: string;
}

type RecvData = {
    phone: string;
    code: string;
}

export const PhoneInput: React.FC<IPhoneInputProps> = props => {

    const { intl, setBusy, verify } = props;

    const [phone, setPhone] = React.useState(`+${props.countryCode} ${props.phoneNo}`);
    const [country, setCountry] = React.useState(props.countryCode);
    const [verified, setVerified] = React.useState(false);
    const [disabled, setDisabled] = React.useState(false);
    const [code, setCode] = React.useState('');
    const [recvCode, setRecvCode] = React.useState<RecvData[]>([]);
    const [codeMsg, setCodeMsg] = React.useState<MessageDescriptor | undefined>(undefined);

    // phone input event handlers
    const phoneChange = (value: string, data: any) => {
        if (data.dialCode === '82') {
            if (value.length > (data.dialCode.length + 2)) {
                if (value[data.dialCode.length + 2] !== '0') {
                    setPhone(value.slice(0, data.dialCode.length + 2) + '0' + value.slice(data.dialCode.length + 2));
                    setCountry(data.dialCode);
                    return;
                }
            }
        }

        setPhone(value);
        setCountry(data.dialCode);
    }

    const sendCode = async () => {
        // let lang = 'en';
        // if (country === '86') lang = 'zh';
        // else if (country === '82') lang = 'ko';
        // else if (country === '852') lang = 'zh-hk';
        // else if (country === '886') lang = 'zh-tw';
        let realNo = phone.slice(country.length + 2).replace(/[^0-9]+/g, '');   // remove all non-digit characters
        if (country === '82') realNo = realNo.replace(/^0+/g, '');              // for korean phone, remove leading '0's

        setBusy(true);
        try {
            const { response } = await MemberApi.verifyPhone(country, realNo);
            console.log(response);
            if (recvCode.some(item => phone === item.phone)) {
                setRecvCode([...recvCode, { phone, code: response }]);
            } else {
                setRecvCode([{ phone, code: response }]);
            }

            setBusy(false);
            setDisabled(true);
            setCodeMsg(undefined);
        } catch (error) {
            console.log('Signup.SendCode: ', error);
            setBusy(false);
            setCodeMsg(undefined);
        }
    }

    const verifyCode = () => {
        let realNo = phone.slice(country.length + 2).replace(/[^0-9]+/g, '');   // remove all non-digit characters
        if (country === '82') realNo = realNo.replace(/^0+/g, '');              // for korean phone, remove leading '0's
        if (recvCode.some(item => (item.phone === phone && item.code === code))) {
            setVerified(true);
            verify(country, realNo, true);
            setCodeMsg(undefined);
        } else {
            setVerified(false);
            verify(country, realNo, false);
            setCodeMsg(authMsgs.phoneMismatch);
        }
    }

    const tryOtherPhone = () => {
        setVerified(false);
        setDisabled(false);
        setCode('');
    }

    const codeChange = e => {
        setCode(e.target.value);
        setCodeMsg(undefined);
    }

    // phone helper messages
    const notVerifiedCodeHelper = handler => (
        <Box id='notverified-code-helper' style={{ display: 'inline', paddingRight: 8 }}>
            {/* {intl.formatMessage(authMsgs.gotCode)}&nbsp; */}
            <Link onClick={handler} style={{ cursor: 'pointer' }}>
                {intl.formatMessage(authMsgs.phoneResend)}
            </Link>
        </Box>
    )

    // phone helper messages
    const codeHelper = verified ? undefined : notVerifiedCodeHelper(sendCode);
    const sendable = phone.length > (country.length + 2);
    const classes = useStyles({});

    return (
        <Box className={classes.root}>
            <Box className={classes && classes.row}>
                <Grid container spacing={1} className={classes && classes.countryCode}>
                    <Grid item xs={12} sm={8}>
                        <MuiPhoneInput
                            onChange={phoneChange}
                            value={phone}
                            defaultCountry='ko'
                            disableAreaCodes
                            variant='outlined'
                            id='phone-input'
                            fullWidth
                            required
                            label={intl.formatMessage(authMsgs.phoneLabel)}
                            inputProps={{ margin: 'dense', style: { padding: '12px 14px' } }}
                            margin='dense'
                            countryCodeEditable={false}
                            disabled={disabled}
                            onlyCountries={SUPPORTED_COUNTRIES}
                        />
                    </Grid>
                    <Grid item xs={12} sm={4}>
                        <Button
                            fullWidth
                            id='send-code-btn'
                            variant="contained"
                            color="primary"
                            className={classes && classes.getcode}
                            disabled={!sendable || disabled}
                            onClick={sendCode}
                        >
                            {intl.formatMessage(authMsgs.getcode)}
                        </Button>
                    </Grid>
                </Grid>
            </Box>
            {disabled && (
                <Box style={{ width: '100%', paddingTop: 4 }}>
                    {!verified && (
                        <Grid container spacing={1}>
                            <Grid item xs={12} sm={8}>
                                <TextField
                                    id='phone-code'
                                    name='phoneCode'
                                    variant='outlined'
                                    label={intl.formatMessage(authMsgs.phoneCode)}
                                    fullWidth
                                    margin='dense'
                                    error={!!codeMsg}
                                    helperText={!!codeMsg && intl.formatMessage(codeMsg)}
                                    value={code}
                                    onChange={codeChange}
                                />
                            </Grid>
                            <Grid item xs={12} sm={4} style={{ paddingTop: 12, paddingBottom: 8 }}>
                                <Button
                                    fullWidth
                                    variant="contained"
                                    color="primary"
                                    id="verify-btn"
                                    className={classes && classes.getcode}
                                    disabled={code.length !== 6}
                                    onClick={verifyCode}
                                >
                                    {intl.formatMessage(authMsgs.phoneVerify)}
                                </Button>
                            </Grid>
                        </Grid>
                    )}
                    {codeHelper}
                    <Link id='try-otherphone-link' onClick={tryOtherPhone} style={{ cursor: 'pointer' }}>
                        {intl.formatMessage(authMsgs.phoneTryOther)}
                    </Link>{''}
                </Box>
            )}
        </Box>
    )
}

export default injectIntl(PhoneInput);