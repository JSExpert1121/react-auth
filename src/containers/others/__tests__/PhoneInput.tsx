import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16/build';
import { act } from 'react-dom/test-utils';

// material ui
import Button from '@material-ui/core/Button';
import Link from '@material-ui/core/Link';
import TextField from '@material-ui/core/TextField';
import { createShallow } from '@material-ui/core/test-utils';

// custom
import MuiPhoneInput from 'material-ui-phone-number';
import { PhoneInput } from '../PhoneInput';
jest.mock('service/member');

configure({ adapter: new Adapter() });

describe('containers/others/PhoneInput', () => {

    const fnVerify = jest.fn();
    const fnBusy = jest.fn();
    const fnIntl = { formatMessage: jest.fn() };
    const phoneNo = '';
    fnIntl.formatMessage.mockReturnValue('Code mismatch')

    const shallow = createShallow();
    const wrapper = shallow(
        <PhoneInput
            verify={fnVerify}
            setBusy={fnBusy}
            intl={fnIntl}
            phoneNo={phoneNo}
            countryCode={'82'}
        />
    );

    test('snapshot test', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('UI components', () => {
        expect(wrapper.find(MuiPhoneInput)).toHaveLength(1);
        const phoneCtrl = wrapper.find(MuiPhoneInput) as any;
        expect(phoneCtrl.props().value).toEqual(`+82 ${phoneNo}`);

        expect(wrapper.find(Button)).toHaveLength(1);
        expect(wrapper.find(TextField)).toHaveLength(0);
        expect(wrapper.find(Link)).toHaveLength(0);
    });

    test('interaction - phone no change - 1', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+82 100', { dialCode: '82' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+82 0100');
            done();
        });
    });

    test('interaction - phone no change - 2', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+82 ', { dialCode: '82' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+82 ');
            done();
        });
    });

    test('interaction - phone no change - 3', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+82 01', { dialCode: '82' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+82 01');
            done();
        });
    });

    test('interaction - phone no change - 4', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+86 171', { dialCode: '86' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+86 171');
            done();
        });
    });

    test('interaction - hit send button - 1', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toEqual(true);
            expect(wrapper.find(Button)).toHaveLength(2);
            expect(wrapper.find(TextField)).toHaveLength(1);
            expect(wrapper.find(Link)).toHaveLength(2);
            done();
        });
    });

    test('interaction - resend code', async done => {
        await wrapper.find(Link).first().props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toBe(true);
            expect(wrapper.find(Button)).toHaveLength(2);
            expect(wrapper.find(Button).first().props().disabled).toBe(true);
            expect(wrapper.find(Button).last().props().disabled).toBe(true);
            expect(wrapper.find(Link)).toHaveLength(2);
            expect(wrapper.find(TextField)).toHaveLength(1);
            done();
        });
    });

    test('interaction - try other phone', done => {
        act(() => {
            wrapper.find(Link).last().simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toBe(false);
            expect(wrapper.find(Button)).toHaveLength(1);
            expect(wrapper.find(Button).props().disabled).toBe(false);
            expect(wrapper.find(Link)).toHaveLength(0);
            expect(wrapper.find(TextField)).toHaveLength(0);
            done();
        });
    });

    test('interaction - phone no change - 5', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+82 100', { dialCode: '82' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+82 0100');
            done();
        });
    });

    test('interaction - hit send button - 2', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toEqual(true);
            expect(wrapper.find(Button)).toHaveLength(2);
            expect(wrapper.find(Button).first().props().disabled).toBe(true);
            expect(wrapper.find(Button).last().props().disabled).toBe(true);
            expect(wrapper.find(TextField)).toHaveLength(1);
            expect(wrapper.find(Link)).toHaveLength(2);
            done();
        });
    });

    test('interaction - code change - 1', done => {
        act(() => {
            wrapper.find(TextField).simulate('change', { target: { value: '1020' } });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Button).last().props().disabled).toBe(true);
            expect(wrapper.find(TextField).props().value).toEqual('1020');
            done();
        });
    });

    test('interaction - code change - length = 6', done => {
        act(() => {
            wrapper.find(TextField).simulate('change', { target: { value: '102040' } });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Button).last().props().disabled).toBe(false);
            expect(wrapper.find(TextField).props().value).toEqual('102040');
            done();
        });
    });

    test('interaction - hit verify button', done => {
        act(() => {
            wrapper.find(Button).last().simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Button).last().props().disabled).toBe(false);
            expect(wrapper.find(TextField).props().value).toEqual('102040');
            expect(fnVerify).lastCalledWith('82', '100', false);
            expect(wrapper.find(TextField).props().helperText).toEqual('Code mismatch');
            expect(wrapper.find(TextField).props().error).toBe(true);
            done();
        });
    });

    test('interaction - code change - correct code', done => {
        act(() => {
            wrapper.find(TextField).simulate('change', { target: { value: '102030' } });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Button).last().props().disabled).toBe(false);
            expect(wrapper.find(TextField).props().value).toEqual('102030');
            done();
        });
    });

    test('interaction - hit verify button - correct code', done => {
        act(() => {
            wrapper.find(Button).last().simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(fnVerify).lastCalledWith('82', '100', true);
            expect(wrapper.find(TextField)).toHaveLength(0);
            done();
        });
    });

    test('interaction - try other phone - 2', done => {
        act(() => {
            wrapper.find(Link).last().simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toBe(false);
            expect(wrapper.find(Button)).toHaveLength(1);
            expect(wrapper.find(Button).props().disabled).toBe(false);
            expect(wrapper.find(Link)).toHaveLength(0);
            expect(wrapper.find(TextField)).toHaveLength(0);
            done();
        });
    });

    test('interaction - phone no change - 6', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+88 100', { dialCode: '88' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+88 100');
            done();
        });
    });

    test('interaction - hit send button - 3', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toBe(false);
            expect(wrapper.find(Button)).toHaveLength(1);
            expect(wrapper.find(Button).props().disabled).toBe(false);
            expect(wrapper.find(Link)).toHaveLength(0);
            expect(wrapper.find(TextField)).toHaveLength(0);
            done();
        });
    });

    test('interaction - phone no change - 6', done => {
        act(() => {
            wrapper.find(MuiPhoneInput).simulate('change', '+86 171', { dialCode: '86' });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().value).toEqual('+86 171');
            done();
        });
    });

    test('interaction - hit send button - 4', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(MuiPhoneInput).props().disabled).toEqual(true);
            expect(wrapper.find(Button)).toHaveLength(2);
            expect(wrapper.find(Button).first().props().disabled).toBe(true);
            expect(wrapper.find(Button).last().props().disabled).toBe(true);
            expect(wrapper.find(TextField)).toHaveLength(1);
            expect(wrapper.find(Link)).toHaveLength(2);
            done();
        });
    });

    test('interaction - code change - correct code - 2', done => {
        act(() => {
            wrapper.find(TextField).simulate('change', { target: { value: '102030' } });
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Button).last().props().disabled).toBe(false);
            expect(wrapper.find(TextField).props().value).toEqual('102030');
            done();
        });
    });

    test('interaction - hit verify button - correct code - 2', done => {
        act(() => {
            wrapper.find(Button).last().simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(fnVerify).lastCalledWith('86', '171', true);
            expect(wrapper.find(TextField)).toHaveLength(0);
            done();
        });
    });
});
