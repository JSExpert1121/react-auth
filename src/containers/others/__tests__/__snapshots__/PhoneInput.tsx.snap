// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/others/PhoneInput snapshot test 1`] = `ShallowWrapper {}`;
